File: docs/eos/examples/compliance-starter/redaction.mjs
// redaction.mjs — strip regulated fields BEFORE any third-party / cross-border / LLM call.
// Zero-dependency, deterministic, offline. This is the *code form* of the Agentic data-boundary
// (docs/checklists/F-compliance.md) and eos-doctor D5, enforcing the rule
// "Never place secrets or PII in prompts or logs. Redact before sending to the provider."
// (.github/instructions/ai/10-ai-llm.instructions.md).
//
// Regime presets (compose only what your /compliance profile selected):
//   createRedactor(['PCI-DSS'])          -> base + cardholder data only
//   createRedactor(['HIPAA'])            -> base + PHI (personal identifiers + health fields)
//   createRedactor(['GDPR','PIPL'])      -> base + personal data
//   createRedactor()                     -> all regimes (the default export below)
//
// Each redactor exposes:
//   redact(value)       -> deep clone with regulated fields masked (safe to send / log)
//   scan(value)         -> array of leak locations (empty == clean)
//   assertClean(value)  -> throws if regulated data would leave the boundary

import { readFileSync } from 'node:fs';

// ── Luhn (card) check so we only redact real PANs, not any 13–19 digit id ──────────────
export function luhnValid(digits) {
  if (!/^\d{13,19}$/.test(digits)) return false;
  let sum = 0;
  let alt = false;
  for (let i = digits.length - 1; i >= 0; i--) {
    let n = digits.charCodeAt(i) - 48;
    if (alt) {
      n *= 2;
      if (n > 9) n -= 9;
    }
    sum += n;
    alt = !alt;
  }
  return sum % 10 === 0;
}

// ── Value patterns (catch PII in free text / unknown keys). `validate` gates a match. ──
const EMAIL = { label: 'email', re: /[\w.+-]+@[\w-]+\.[\w.-]+/ };
const SSN = { label: 'ssn', re: /\b\d{3}-\d{2}-\d{4}\b/ };
const PHONE = { label: 'phone', re: /\b\d{3}[-.\s]?\d{3}[-.\s]?\d{4}\b/ };
const PAN = { label: 'pan', re: /\b(?:\d[ -]?){13,19}\b/, validate: (m) => luhnValid(m.replace(/\D/g, '')) };

// Shared "direct personal identifiers" — HIPAA (18 Safe Harbor items) and GDPR/PIPL both want these out.
const PERSONAL_KEYS = ['email', 'phone', 'telephone', 'fax', 'dob', 'dateofbirth', 'birthdate', 'address', 'ssn', 'socialsecurity'];
const PERSONAL_PATTERNS = [EMAIL, SSN, PHONE];

// ── Regime profiles. keys = case-insensitive SUBSTRING match on object keys. ───────────
// ⚠ Substring matching means SHORT tokens over-match ('pin'⊂'shipping', 'track'⊂'tracking',
//   'sin'⊂'business', 'ip'⊂'description'). Tokens below are curated to be distinctive; if your
//   schema still collides, switch isDeniedKey to exact-match or scope the key list per module.
export const PROFILES = {
  // Credentials/secrets — never leave the boundary regardless of regulatory regime.
  base: {
    keys: ['password', 'passwd', 'secret', 'token', 'apikey', 'api_key', 'accesskey', 'credential', 'privatekey', 'private_key'],
    patterns: [],
  },
  // HIPAA: PHI = personal identifiers + health-specific fields (Security/Privacy Rule, Safe Harbor).
  HIPAA: {
    keys: [...PERSONAL_KEYS, 'mrn', 'medicalrecord', 'medical_record', 'phi', 'diagnosis', 'icd', 'healthplan', 'health_plan', 'beneficiary', 'npi', 'dea', 'patientid'],
    patterns: [...PERSONAL_PATTERNS],
  },
  // PCI-DSS: cardholder data (CHD) + sensitive authentication data (SAD) only — scoped to cards.
  PCI: {
    keys: ['pan', 'cardnumber', 'card_number', 'cardno', 'cvv', 'cvc', 'cvv2', 'track1', 'track2', 'magstripe', 'pinblock', 'expiry', 'expiration'],
    patterns: [PAN],
  },
  // GDPR / CCPA / PIPL: personal data — identifiers + national/sensitive PI.
  GDPR_PIPL: {
    keys: [...PERSONAL_KEYS, 'nationalid', 'national_id', 'idcard', 'passport', 'taxid', 'drivinglicense', 'driverslicense', 'biometric', 'geolocation'],
    patterns: [...PERSONAL_PATTERNS],
  },
};

// Map the regime names /compliance uses onto profile keys.
export const REGIME_ALIASES = {
  hipaa: 'HIPAA',
  pci: 'PCI', 'pci-dss': 'PCI', pcidss: 'PCI',
  gdpr: 'GDPR_PIPL', ccpa: 'GDPR_PIPL', 'ccpa-cpra': 'GDPR_PIPL', pipl: 'GDPR_PIPL',
};

const DEFAULT_REGIMES = ['HIPAA', 'PCI', 'GDPR_PIPL'];
const MASK = (label) => `«REDACTED:${label}»`;
const isMask = (v) => typeof v === 'string' && /^«REDACTED:[^»]+»$/.test(v);
const isEmpty = (v) => v == null || v === '';

function toProfileName(r) {
  if (PROFILES[r]) return r;
  return REGIME_ALIASES[String(r).toLowerCase()] || null;
}

// Compose base + selected regimes into a deduped { keys, patterns } config.
function resolveProfiles(regimes) {
  const names = ['base'];
  for (const r of regimes || []) {
    const n = toProfileName(r);
    if (n && !names.includes(n)) names.push(n);
  }
  const keys = [];
  const patterns = [];
  const seenKey = new Set();
  const seenPat = new Set();
  for (const n of names) {
    const p = PROFILES[n];
    if (!p) continue;
    for (const k of p.keys) {
      const lk = k.toLowerCase();
      if (!seenKey.has(lk)) { seenKey.add(lk); keys.push(lk); }
    }
    for (const pat of p.patterns) {
      if (!seenPat.has(pat.label)) { seenPat.add(pat.label); patterns.push(pat); }
    }
  }
  // Validate-gated patterns (PAN) first — they consume digit runs before generic matchers.
  patterns.sort((a, b) => (b.validate ? 1 : 0) - (a.validate ? 1 : 0));
  return { active: names.filter((n) => n !== 'base'), keys, patterns };
}

const isDeniedKey = (k, keys) => { const lk = k.toLowerCase(); return keys.some((d) => lk.includes(d)); };

function redactString(s, patterns) {
  let out = s;
  for (const { label, re, validate } of patterns) {
    out = out.replace(new RegExp(re.source, 'g'), (m) => (validate ? (validate(m) ? MASK(label) : m) : MASK(label)));
  }
  return out;
}

function redactValue(value, cfg) {
  if (Array.isArray(value)) return value.map((x) => redactValue(x, cfg));
  if (value && typeof value === 'object') {
    const out = {};
    for (const [k, v] of Object.entries(value)) out[k] = isDeniedKey(k, cfg.keys) ? MASK(k.toLowerCase()) : redactValue(v, cfg);
    return out;
  }
  if (typeof value === 'string') return redactString(value, cfg.patterns);
  return value;
}

function scanValue(value, cfg) {
  const hits = [];
  const walk = (v, p) => {
    if (Array.isArray(v)) v.forEach((x, i) => walk(x, `${p}[${i}]`));
    else if (v && typeof v === 'object')
      for (const [k, val] of Object.entries(v)) {
        if (isDeniedKey(k, cfg.keys) && !isEmpty(val) && !isMask(val)) hits.push({ path: `${p}.${k}`, kind: `key:${k.toLowerCase()}` });
        walk(val, `${p}.${k}`);
      }
    else if (typeof v === 'string' && redactString(v, cfg.patterns) !== v) hits.push({ path: p, kind: 'value-pattern' });
  };
  walk(value, '$');
  return hits;
}

// Build a redactor bound to the selected regimes. This is the main entry point.
export function createRedactor(regimes = DEFAULT_REGIMES) {
  const cfg = resolveProfiles(regimes);
  const assertClean = (value) => {
    const hits = scanValue(value, cfg);
    if (hits.length) {
      const e = new Error(
        `redaction: ${hits.length} regulated field(s) would leave the boundary: ` +
          hits.map((h) => `${h.path} (${h.kind})`).join(', '),
      );
      e.violations = hits;
      throw e;
    }
    return value;
  };
  return {
    regimes: cfg.active,
    keys: cfg.keys,
    patterns: cfg.patterns.map((p) => p.label),
    redact: (v) => redactValue(v, cfg),
    redactForLog: (v) => redactValue(v, cfg), // PII-free logging alias
    scan: (v) => scanValue(v, cfg),
    assertClean,
  };
}

// ── Auto-select the regime(s) from the /compliance output ──────────────────────────────
// `/compliance` writes docs/compliance-profile.md with a canonical machine-readable line:
//   **Regulatory regime:** HIPAA, PCI-DSS      (or `none` for generic PII handling)
// parseRegimes reads that line and resolves it to profile names. Only tokens that resolve
// to a *known* regime are kept, so free-text rationale on the line is ignored; an explicit
// `none` short-circuits to [] (base credentials only).
export function parseRegimes(profileText) {
  const m = /regulatory\s+regime\s*[:：]\s*(.+)/i.exec(String(profileText || ''));
  if (!m) return [];
  let val = m[1].replace(/^[*_\s]+/, ''); // drop leading markdown emphasis (**bold**)
  val = val.split(/[(（]/)[0]; // drop parenthetical rationale before tokenizing
  if (/^none\b/i.test(val.trim())) return [];
  const out = [];
  for (const tok of val.split(/[\s,/+&]+/)) {
    if (!tok) continue;
    const n = toProfileName(tok);
    if (n && n !== 'base' && !out.includes(n)) out.push(n);
  }
  return out;
}

// Build a redactor straight from the /compliance profile doc.
//   file lists regimes  -> scope to them
//   file says `none`     -> base credentials only (honours the human decision)
//   file missing         -> fallback: 'all' (default, fail-safe — never silently under-redact),
//                           'base', or 'throw'
export function redactorFromProfile(path = 'docs/compliance-profile.md', { fallback = 'all' } = {}) {
  let text;
  try {
    text = readFileSync(path, 'utf8');
  } catch {
    if (fallback === 'throw') {
      throw new Error(`redactorFromProfile: ${path} not found — run /compliance first`);
    }
    return createRedactor(fallback === 'base' ? [] : undefined); // 'all' -> default regimes
  }
  return createRedactor(parseRegimes(text));
}

// ── Default instance = all regimes. Back-compat top-level exports. ─────────────────────
const _default = createRedactor();
export const redact = _default.redact;
export const redactForLog = _default.redactForLog;
export const scan = _default.scan;
export const assertClean = _default.assertClean;
// Composed deny list of the default (all-regime) instance — for inspection / back-compat.
export const DENY_KEYS = _default.keys;

File: docs/eos/examples/compliance-starter/redaction.py
"""redaction.py — strip regulated fields BEFORE any third-party / cross-border / LLM call.

Zero-dependency (stdlib only), deterministic, offline. Python parallel of ``redaction.mjs``.
This is the *code form* of the Agentic data-boundary (docs/checklists/F-compliance.md) and
eos-doctor D5, enforcing the rule "Never place secrets or PII in prompts or logs. Redact
before sending to the provider." (.github/instructions/ai/10-ai-llm.instructions.md).

Regime presets (compose only what your /compliance profile selected)::

    create_redactor(["PCI-DSS"])         # base + cardholder data only
    create_redactor(["HIPAA"])           # base + PHI (personal identifiers + health fields)
    create_redactor(["GDPR", "PIPL"])    # base + personal data
    create_redactor()                    # all regimes (also the module-level default below)

A redactor exposes ``redact(value)``, ``scan(value)``, ``assert_clean(value)``.
"""
from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Callable, Iterable, NamedTuple


# ── Luhn (card) check so we only redact real PANs, not any 13–19 digit id ──────────────
def luhn_valid(digits: str) -> bool:
    if not re.fullmatch(r"\d{13,19}", digits or ""):
        return False
    total, alt = 0, False
    for ch in reversed(digits):
        n = ord(ch) - 48
        if alt:
            n *= 2
            if n > 9:
                n -= 9
        total += n
        alt = not alt
    return total % 10 == 0


# ── Value patterns (catch PII in free text / unknown keys). ``validate`` gates a match. ─
class Pattern(NamedTuple):
    label: str
    regex: re.Pattern
    validate: Callable[[str], bool] | None = None


EMAIL = Pattern("email", re.compile(r"[\w.+-]+@[\w-]+\.[\w.-]+"))
SSN = Pattern("ssn", re.compile(r"\b\d{3}-\d{2}-\d{4}\b"))
PHONE = Pattern("phone", re.compile(r"\b\d{3}[-.\s]?\d{3}[-.\s]?\d{4}\b"))
PAN = Pattern("pan", re.compile(r"\b(?:\d[ -]?){13,19}\b"), lambda m: luhn_valid(re.sub(r"\D", "", m)))

# Shared "direct personal identifiers" — HIPAA (18 Safe Harbor items) and GDPR/PIPL both want these out.
PERSONAL_KEYS = ["email", "phone", "telephone", "fax", "dob", "dateofbirth", "birthdate",
                 "address", "ssn", "socialsecurity"]
PERSONAL_PATTERNS = [EMAIL, SSN, PHONE]


@dataclass(frozen=True)
class Profile:
    keys: list[str]
    patterns: list[Pattern]


# ── Regime profiles. keys = case-insensitive SUBSTRING match on mapping keys. ───────────
# ⚠ Substring matching means SHORT tokens over-match ('pin'⊂'shipping', 'track'⊂'tracking',
#   'sin'⊂'business'). Tokens below are curated to be distinctive; if your schema still
#   collides, switch _is_denied_key to exact-match or scope the key list per module.
PROFILES: dict[str, Profile] = {
    # Credentials/secrets — never leave the boundary regardless of regulatory regime.
    "base": Profile(
        keys=["password", "passwd", "secret", "token", "apikey", "api_key", "accesskey",
              "credential", "privatekey", "private_key"],
        patterns=[],
    ),
    # HIPAA: PHI = personal identifiers + health-specific fields (Security/Privacy Rule, Safe Harbor).
    "HIPAA": Profile(
        keys=[*PERSONAL_KEYS, "mrn", "medicalrecord", "medical_record", "phi", "diagnosis",
              "icd", "healthplan", "health_plan", "beneficiary", "npi", "dea", "patientid"],
        patterns=[*PERSONAL_PATTERNS],
    ),
    # PCI-DSS: cardholder data (CHD) + sensitive authentication data (SAD) only — scoped to cards.
    "PCI": Profile(
        keys=["pan", "cardnumber", "card_number", "cardno", "cvv", "cvc", "cvv2", "track1",
              "track2", "magstripe", "pinblock", "expiry", "expiration"],
        patterns=[PAN],
    ),
    # GDPR / CCPA / PIPL: personal data — identifiers + national/sensitive PI.
    "GDPR_PIPL": Profile(
        keys=[*PERSONAL_KEYS, "nationalid", "national_id", "idcard", "passport", "taxid",
              "drivinglicense", "driverslicense", "biometric", "geolocation"],
        patterns=[*PERSONAL_PATTERNS],
    ),
}

# Map the regime names /compliance uses onto profile keys.
REGIME_ALIASES = {
    "hipaa": "HIPAA",
    "pci": "PCI", "pci-dss": "PCI", "pcidss": "PCI",
    "gdpr": "GDPR_PIPL", "ccpa": "GDPR_PIPL", "ccpa-cpra": "GDPR_PIPL", "pipl": "GDPR_PIPL",
}

_DEFAULT_REGIMES = ["HIPAA", "PCI", "GDPR_PIPL"]
_MASK_RE = re.compile(r"^«REDACTED:[^»]+»$")


def _mask(label: str) -> str:
    return f"«REDACTED:{label}»"


def _is_mask(v: Any) -> bool:
    return isinstance(v, str) and bool(_MASK_RE.match(v))


def _is_empty(v: Any) -> bool:
    return v is None or v == ""


def _to_profile_name(regime: str) -> str | None:
    if regime in PROFILES:
        return regime
    return REGIME_ALIASES.get(str(regime).lower())


class Redactor:
    """Redactor bound to a set of regimes. Build via ``create_redactor``."""

    def __init__(self, regimes: Iterable[str] | None = None):
        names = ["base"]
        for r in (regimes if regimes is not None else _DEFAULT_REGIMES):
            n = _to_profile_name(r)
            if n and n not in names:
                names.append(n)
        keys: list[str] = []
        patterns: list[Pattern] = []
        seen_key: set[str] = set()
        seen_pat: set[str] = set()
        for n in names:
            prof = PROFILES.get(n)
            if not prof:
                continue
            for k in prof.keys:
                lk = k.lower()
                if lk not in seen_key:
                    seen_key.add(lk)
                    keys.append(lk)
            for p in prof.patterns:
                if p.label not in seen_pat:
                    seen_pat.add(p.label)
                    patterns.append(p)
        # Validate-gated patterns (PAN) first — they consume digit runs before generic matchers.
        patterns.sort(key=lambda p: 0 if p.validate else 1)
        self.regimes = [n for n in names if n != "base"]
        self.keys = keys
        self.patterns = patterns

    # ── internals ──────────────────────────────────────────────────────────────────────
    def _is_denied_key(self, k: str) -> bool:
        lk = k.lower()
        return any(d in lk for d in self.keys)

    def _redact_string(self, s: str) -> str:
        out = s
        for p in self.patterns:
            if p.validate is None:
                out = p.regex.sub(_mask(p.label), out)
            else:
                out = p.regex.sub(lambda m, _p=p: _mask(_p.label) if _p.validate(m.group(0)) else m.group(0), out)
        return out

    # ── public API ───────────────────────────────────────────────────────────────────--
    def redact(self, value: Any) -> Any:
        """Deep copy with regulated fields masked. Safe to send to a provider or log."""
        if isinstance(value, list):
            return [self.redact(x) for x in value]
        if isinstance(value, dict):
            return {k: (_mask(k.lower()) if self._is_denied_key(k) else self.redact(v)) for k, v in value.items()}
        if isinstance(value, str):
            return self._redact_string(value)
        return value

    # PII-free logging alias — data-api: "every deletion audited … without logging the data itself".
    def redact_for_log(self, value: Any) -> Any:
        return self.redact(value)

    def scan(self, value: Any) -> list[dict]:
        """Report every place regulated data would leak (path + kind). Empty list == clean."""
        hits: list[dict] = []

        def walk(v: Any, p: str) -> None:
            if isinstance(v, list):
                for i, x in enumerate(v):
                    walk(x, f"{p}[{i}]")
            elif isinstance(v, dict):
                for k, val in v.items():
                    if self._is_denied_key(k) and not _is_empty(val) and not _is_mask(val):
                        hits.append({"path": f"{p}.{k}", "kind": f"key:{k.lower()}"})
                    walk(val, f"{p}.{k}")
            elif isinstance(v, str) and self._redact_string(v) != v:
                hits.append({"path": p, "kind": "value-pattern"})

        walk(value, "$")
        return hits

    def assert_clean(self, value: Any) -> Any:
        """Boundary guard — call right before any cross-border / third-party / LLM request."""
        hits = self.scan(value)
        if hits:
            detail = ", ".join(f"{h['path']} ({h['kind']})" for h in hits)
            err = ValueError(f"redaction: {len(hits)} regulated field(s) would leave the boundary: {detail}")
            err.violations = hits  # type: ignore[attr-defined]
            raise err
        return value


def create_redactor(regimes: Iterable[str] | None = None) -> Redactor:
    """Build a redactor bound to the selected regimes. Main entry point."""
    return Redactor(regimes)


# ── Auto-select the regime(s) from the /compliance output ──────────────────────────────
# ``/compliance`` writes docs/compliance-profile.md with a canonical machine-readable line::
#
#     **Regulatory regime:** HIPAA, PCI-DSS      (or ``none`` for generic PII handling)
#
# ``parse_regimes`` reads that line and resolves it to profile names. Only tokens that
# resolve to a *known* regime are kept, so free-text rationale on the line is ignored; an
# explicit ``none`` short-circuits to ``[]`` (base credentials only).
_REGIME_LINE = re.compile(r"regulatory\s+regime\s*[:：]\s*(.+)", re.IGNORECASE)
_REGIME_SPLIT = re.compile(r"[\s,/+&]+")


def parse_regimes(profile_text: str) -> list[str]:
    """Extract resolved regime names from a /compliance profile's regime line."""
    m = _REGIME_LINE.search(profile_text or "")
    if not m:
        return []
    val = re.sub(r"^[*_\s]+", "", m.group(1))  # drop leading markdown emphasis (**bold**)
    val = re.split(r"[(（]", val)[0]  # drop parenthetical rationale before tokenizing
    if re.match(r"none\b", val.strip(), re.IGNORECASE):
        return []
    out: list[str] = []
    for tok in _REGIME_SPLIT.split(val):
        if not tok:
            continue
        n = _to_profile_name(tok)
        if n and n != "base" and n not in out:
            out.append(n)
    return out


def redactor_from_profile(
    path: str | Path = "docs/compliance-profile.md", *, fallback: str = "all"
) -> Redactor:
    """Build a redactor straight from the /compliance profile doc.

    * file lists regimes -> scope to them
    * file says ``none``  -> base credentials only (honours the human decision)
    * file missing        -> ``fallback``: ``"all"`` (default, fail-safe — never silently
      under-redact), ``"base"``, or ``"throw"``.
    """
    try:
        text = Path(path).read_text(encoding="utf-8")
    except FileNotFoundError:
        if fallback == "throw":
            raise FileNotFoundError(
                f"redactor_from_profile: {path} not found — run /compliance first"
            ) from None
        return create_redactor([] if fallback == "base" else None)
    return create_redactor(parse_regimes(text))


# ── Default instance = all regimes. Module-level convenience API. ──────────────────────
_default = create_redactor()
redact = _default.redact
redact_for_log = _default.redact_for_log
scan = _default.scan
assert_clean = _default.assert_clean
DENY_KEYS = _default.keys  # composed deny list of the default (all-regime) instance
